src/mongo_odm/expression_syntax.hpp

```diff
@@ -0,0 +1,176 @@
+// Copyright 2016 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <mongo_odm/config/prelude.hpp>
+
+#include <cstddef>
+#include <stdexcept>
+#include <tuple>
+#include <type_traits>
+
+#include <bsoncxx/types.hpp>
+
+#include <mongo_odm/macros.hpp>
+#include <mongo_odm/util.hpp>
+
+namespace mongo_odm {
+MONGO_ODM_INLINE_NAMESPACE_BEGIN
+
+// Enum of possible expression types.
+enum struct expression_category { none, query, update, sort };
+
+// Forward declarations for Expression type trait structs
+template <expression_category, typename>
+struct is_expression_type;
+
+/* Type traits structs for various expressions */
+template <expression_category expr_type>
+struct expression_category_type {
+    static constexpr expression_category value = expr_type;
+};
+
+/* forward declarations of expression classes */
+
+template <typename NvpT>
+class SortExpr;
```

Comment:
It can also be caught in our grand renaming sweep, but I think it's better to do here if possible.